### PR TITLE
Refactor tx lib

### PR DESCRIPTION
**Pull request by ryanio**
This PR continues the efforts from https://github.com/ethereumjs/ethereumjs-tx/pull/154, applying it to the monorepo.

**TL;DR:** For a full description of all the changes, please see the [tx changelog](https://github.com/ethereumjs/ethereumjs-vm/blob/4491218142421cd63412532da92499e2717ba26d/packages/tx/CHANGELOG.md).

This PR:

* Refactors and simplifies the tx library to remove the use of `ethereumjs-util`'s `defineProperties` (see https://github.com/ethereumjs/organization/issues/56)
* Uses types closer to their domain representation (BN, Address, etc.) instead of all `Buffer`s
* Simplifies the constructor params, and
  * adds new static factory helpers `fromTxData`, `fromRlpSerializedTx`, and `fromValuesArray`
  * only accepts `Common`, instead of string options for chain and hardfork

This PR also removes FakeTransaction as I found it burdensome to maintain in this refactor due to `from` not being stored anymore. Its only use was in block's `from-rpc` and can achieve the same result on the normal Transaction class with the same code there, by overriding `getSenderAddress` and `hash`, see [here](https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/block/src/from-rpc.ts#L41) and the note in the [readme](https://github.com/ethereumjs/ethereumjs-vm/blob/4491218142421cd63412532da92499e2717ba26d/packages/tx/README.md#fake-transaction).

---

#### Examples of updating code

> `tx.from`
> &nbsp; =>
> `tx.getSenderAddress()`

> `tx.hash(false)`
> &nbsp; =>
> `tx.getMessageToSign()`

> `new Transaction()`
> &nbsp; =>
> `Transaction.fromTxData({})`

> `new Transaction(tx.rlp)`
> &nbsp; =>
> `Transaction.fromRlpSerializedTx(toBuffer(tx.rlp))`

> `new Transaction(txFixtures[0].raw)`
> &nbsp; =>
> `Transaction.fromValuesArray(txFixtures[0].raw.map(toBuffer))`

> ```
> const s1 = tx.serialize().toString('hex')
> 
> const tx = new Transaction(s1)
>  =>
> const tx = Transaction.fromRlpSerializedTx(toBuffer('0x' + s1))
> ```

Initializing with object:

> ```
> const txData = {
>   gasPrice: 1000,
>   gasLimit: 10000000,
>   value: 42,
> }
> const tx = new Transaction(txData)
> ```
> &nbsp; =>
> ```
> const tx = Transaction.fromTxData(txData)
> ```

### CODE REVIEW

**Comment by holgerd77** [comment]
Love this.

**Comment by ryanio** [comment]
@alcuadrado would you like me to take this one step further with your implementation example in https://github.com/ethereumjs/organization/issues/56?

there are some additional ideas there not implemented in this PR yet:
1. domain types closer to actual representation (rather than buffer for all)
2. freezing the object after construction

**Comment by alcuadrado** [comment]
Great PR! 

> * domain types closer to actual representation (rather than buffer for all)

I think this would be great. It would mean removing `defineProperties` :)

**Comment by ryanio** [comment]
@s1na would you have any idea why this test is now failing with "revert" instead of "value overflow"? the test above it seems fine, so it's not immediately obvious to me, and i didn't see any "revert reason" on the exception or error object to help point toward where it might have failed 🤔

[ci link](https://github.com/ethereumjs/ethereumjs-vm/pull/812/checks?check_run_id=867140782#step:7:343)

```
# should fail when account balance overflows (call)
ok 199 should be equal
ok 200 should be equal
# should fail when account balance overflows (create)
not ok 201 should be equal
  ---
    operator: equal
    expected: 'value overflow'
    actual:   'revert'
    at: Test.tape (/Users/ryanghods/dev/ethereumjs-vm/packages/vm/tests/api/runTx.js:112:5)
    stack: |-
      Error: should be equal
          at Test.assert [as _assert] (/Users/ryanghods/dev/ethereumjs-vm/node_modules/tape/lib/test.js:228:54)
          at Test.bound [as _assert] (/Users/ryanghods/dev/ethereumjs-vm/node_modules/tape/lib/test.js:80:32)
          at Test.equal (/Users/ryanghods/dev/ethereumjs-vm/node_modules/tape/lib/test.js:389:10)
          at Test.bound [as equal] (/Users/ryanghods/dev/ethereumjs-vm/node_modules/tape/lib/test.js:80:32)
          at Test.tape (/Users/ryanghods/dev/ethereumjs-vm/packages/vm/tests/api/runTx.js:112:5)
  ...
```

**Comment by s1na** [comment]
@ryanio hm it doesn't immediately pop into my head why a `call` should work as expected but a `create` should fail. I'd try to find where the revert is happening. You might find a lead if you print the stack trace here https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/vm/lib/evm/evm.ts#L132

File: packages/block/src/block.ts [hunk 1]
```diff
@@ -73,10 +73,29 @@ export class Block {
     }
 
     // parse transactions
+
+    // get common based on block's number
+    // since hardfork param may not be set
+    // (see TODO above)
+    let common = this._common
```
Comment (alcuadrado) [inline packages/block/src/block.ts:80]: Why is this required now? Is the hardfork needed by the tx now?

File: packages/block/src/block.ts [hunk 2]
```diff
@@ -73,10 +73,29 @@ export class Block {
...
+    if (!common.hardfork()) {
+      const blockNumber = parseInt(this.header.number.toString('hex'), 16)
+      const chainId = this._common.chainId()
+      if (blockNumber) {
+        const hfs = this._common.activeHardforks(blockNumber)
+        const hf = hfs[hfs.length - 1].name
+        common = new Common(chainId, hf)
+      } else {
+        // fall back to default hf
+        common = new Common(chainId, 'petersburg')
```
Comment (alcuadrado) [inline packages/block/src/block.ts:90]: Maybe this should be a constant in `Common`

Comment (alcuadrado) [reply packages/block/src/block.ts:90]: `'petersburg'`, I mean

File: packages/block/src/block.ts (hunk 1 above)
Comment (alcuadrado) [reply packages/block/src/block.ts:80]: Self answer: yes

File: packages/block/src/block.ts [hunk 3]
```diff
@@ -73,10 +73,29 @@ export class Block {
...
+      }
+    }
+
     for (let i = 0; i < rawTransactions.length; i++) {
-      // TODO: Pass the common object instead of the options. It can't be implemented right now
-      // because the hardfork may be `null`. Read the above TODO for more info.
-      const tx = new Transaction(rawTransactions[i], chainOptions as TransactionOptions)
+      const txData = rawTransactions[i]
+      const tx = Array.isArray(txData)
+        ? Transaction.fromValuesArray(txData as Buffer[], common)
```
Comment (alcuadrado) [inline packages/block/src/block.ts:97]: Nice! These separate constructors are much better.

File: packages/block/src/from-rpc.ts [hunk 4]
```diff
@@ -32,19 +33,30 @@ export default function blockFromRpc(
...
-      const fromAddress = toBuffer(txParams.from)
+      const fromAddress = new Address(toBuffer(txParams.from))
       delete txParams.from
 
-      const tx = new FakeTransaction(txParams, chainOptions as TransactionOptions)
-      tx.from = fromAddress
+      // get common based on block's number
+      // since hardfork param may not be set
+      // (see TODO in block.constructor)
+      const blockNumber = parseInt(block.header.number.toString('hex'), 16)
```
Comment (alcuadrado) [inline packages/block/src/from-rpc.ts:44]: As this is done multiple times, maybe it should belong to `Common`?

File: packages/tx/src/fake.ts [hunk 5]
```diff
@@ -1,62 +0,0 @@
...
- *
- * @param data - A transaction can be initialized with its rlp representation, an array containing
- * the value of its fields in order, or an object containing them by name.
- *
- * @param opts - The transaction's options, used to indicate the chain and hardfork the
- * transactions belongs to.
- *
- * @see Transaction
- */
-export default class FakeTransaction extends Transaction {
```
Comment (alcuadrado) [inline packages/tx/src/fake.ts:17]: This `FakeTransaction` class is the only way of implementing `eth_call` using `ethereumjs-vm` now, because of this issue: https://github.com/ethereumjs/ethereumjs-vm/issues/613

This is implemented using `FakeTransaction` in both Ganache and Buidler EVM.

I think it's great dropping this here, but I think we should document how `eth_call` can be implemented. Maybe it can be an example in the VM package.

Comment (ryanio) [reply packages/tx/src/fake.ts:17]: Thanks, that sounds good I like the idea of an adding an example of this.

An implementation of `FakeTransaction` used in `runTx` would look something like this (just writing here):

```javascript
import { Transaction, Address } from '@ethereumjs/tx'

public async _getFakeTransaction(
    txParams: TransactionParams
  ): Promise<Transaction> {
  const from = new Address(txParams.from)
  delete txParams.from

  const tx = Transaction.fromTxData(txParams, this._common)

  const fakeTx = Object.create(tx)

  // override getSenderAddress
  fakeTx.getSenderAddress = function () { return from }

  return fakeTx
}
```

File: packages/block/src/block.ts (hunk 2 above)
Comment (ryanio) [reply packages/block/src/block.ts:90]: yeah, maybe I can do in a follow-up PR since it's a default assumption across the broader monorepo like in the StateManager [here][1].

Can we export constants with common? It would be nice if we could import e.g. `DEFAULT_HARDFORK`. I know we also have have #702 that would also be great to get to.

[1]: https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/vm/lib/state/stateManager.ts#L46

File: packages/block/src/from-rpc.ts (hunk 4 above)
Comment (ryanio) [reply packages/block/src/from-rpc.ts:44]: oh right, sure, I can do it as a static factory `Common.forBlockNumber(chain, blockNumber): Common` 

I also considered something like a helper `Common.hardforkForBlockNumber(chain, blockNumber): string` but I think the static factory is a nice api pattern that we can continue to use. And if really needed you could just access the hardfork with `common.hardfork()`.

File: packages/tx/src/transaction.ts [hunk 6]
```diff
@@ -21,290 +21,255 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
...
+
+    return this.fromValuesArray(values, common)
+  }
+
+  public static fromValuesArray(values: Buffer[], common?: Common) {
+    if (values.length > 9) {
+      throw new Error('Invalid transaction. More values than expected were received')
+    }
+
+    const [nonce, gasPrice, gasLimit, to, value, data, v, r, s] = values
```
Comment (alcuadrado) [inline packages/tx/src/transaction.ts:67]: What if the length is `< 9`? 

Maybe `9` and `6` should be the only valid lengths. Like signed and unsigned txs.

File: packages/tx/src/transaction.ts [hunk 7]
```diff
@@ -21,290 +21,255 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
...
+    r?: BN,
+    s?: BN,
   ) {
-    // instantiate Common class instance based on passed options
-    if (opts.common) {
-      if (opts.chain || opts.hardfork) {
-        throw new Error(
-          'Instantiation with both opts.common, and opts.chain and opts.hardfork parameter not allowed!',
-        )
+    if (!(data instanceof Buffer)) {
```
Comment (alcuadrado) [inline packages/tx/src/transaction.ts:100]: I'm not sure if I understand why this is here. I mean, why is only `data` validated? I'd be ok with no type being validated at runtime.

File: packages/tx/src/transaction.ts [hunk 8]
```diff
@@ -21,290 +21,255 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
...
   getDataFee(): BN {
-    const data = this.raw[5]
     const cost = new BN(0)
-    for (let i = 0; i < data.length; i++) {
-      data[i] === 0
-        ? cost.iaddn(this._common.param('gasPrices', 'txDataZero'))
-        : cost.iaddn(this._common.param('gasPrices', 'txDataNonZero'))
+    for (let i = 0; i < this.data.length; i++) {
+      this.data[i] === 0
+        ? cost.iaddn(this.common.param('gasPrices', 'txDataZero'))
```
Comment (alcuadrado) [inline packages/tx/src/transaction.ts:251]: Moving the param outside of the loop would be great. I believe is one of the performance hotspots that ethworks identified.

Also, can this sum be implemented on native numbers and then converted to BNs? If we assume that the `MAX_DATA_LENGTH * txDataNonZero` is < 2**53 that would be fine.

File: packages/tx/src/transaction.ts [hunk 9]
```diff
@@ -315,48 +280,114 @@ export default class Transaction {
...
-      errors.push([`gas limit is too low. Need at least ${this.getBaseFee()}`])
+    if (this.getBaseFee().gt(this.gasLimit)) {
+      errors.push(`gasLimit is too low. given ${this.gasLimit}, need at least ${this.getBaseFee()}`)
     }
 
-    if (stringError === false) {
-      return errors.length === 0
-    } else {
+    if (stringError) {
       return errors.join(' ')
```
Comment (alcuadrado) [inline packages/tx/src/transaction.ts:293]: This `join(' ')` has always looked out of place to me. I understand that `validate(true)` gives more info than just `validate()`, but why then combine that into a single string with an arbitrary format?

If we are introducing breaking changes here, we may want to change this too. Maybe returning `string[]`?

**Comment by alcuadrado** [comment]
I really like how this PR is looking! I haven't yet reviewed the tests, but the new `Transaction` class looks great.

I think we can also take advantage of the immutable interface to more aggressively cache some common operations. When profiling the VM I've seen thinks like hashing, and validating txs taking a considerable part of the execution time. We can have a constant public interface, but keep some mutable/lazily-initialized private properties to cache those.

File: packages/block/src/from-rpc.ts (hunk 4 above)
Comment (ryanio) [reply packages/block/src/from-rpc.ts:44]: added in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/e425377791fe3b5da750bbda520077ab27ff8567

File: packages/tx/src/transaction.ts (hunk 6 above)
Comment (ryanio) [reply packages/tx/src/transaction.ts:68]: good idea, added in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/d6cc6bfb89275bdfb5ac498db0960ff12cd23b52

File: packages/tx/src/transaction.ts (hunk 7 above)
Comment (ryanio) [reply packages/tx/src/transaction.ts]: done in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/20fd30a3481cfa7d5af94ef70bacc7b9869ec854

File: packages/tx/src/transaction.ts (hunk 8 above)
Comment (ryanio) [reply packages/tx/src/transaction.ts]: nice idea, done in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/34bd63942cf766d9cf4ea5d649c27e613ee73d58

File: packages/tx/src/transaction.ts (hunk 9 above)
Comment (ryanio) [reply packages/tx/src/transaction.ts]: 👍 agree, done in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/420d20ba988d3f5d2e4cf9317520fb981d2a038a

When I work on `block` I will have to remember to do the same to `block.validateTransactions(stringError)`

**Comment by holgerd77** [comment]
Just did a short recap on reasoning and outline from Patricio's initial PR and skimmed through the code here. Not enough for a review, but at least can confirm that I am fully supportive on the changes done in the PR.

File: packages/block/src/block.ts (hunk 2 above)
Comment (ryanio) [reply packages/block/src/block.ts:88]: opened an issue for this https://github.com/ethereumjs/ethereumjs-vm/issues/822

File: packages/block/src/from-rpc.ts (hunk 4 above)
Comment (ryanio) [reply packages/block/src/from-rpc.ts:44]: I may remove this in favor of `Common.setHardforkByBlockNumber()` which is being added in https://github.com/ethereumjs/ethereumjs-vm/pull/863.

**Comment by holgerd77** [comment]
Can this be updated so that it can get its deserved review? 😄

**Comment by holgerd77** [comment]
Hi Ryan, already mentioned your work here in two dedicated issues, can you eventually generally go through the [issues opened on the tx library](https://github.com/ethereumjs/ethereumjs-vm/issues?q=is%3Aopen+is%3Aissue+label%3A%22package%3A+tx%22) along your work here and see if some of these are already (or could be easily) addressed and therefore closed upon merge? That would be great! 😄

**Comment by ryanio** [comment]
This PR should resolve the following issues: #704, #709,  #710, #711, #712, #714, #829

File: packages/block/src/block.ts [hunk 10]
```diff
@@ -63,8 +63,15 @@ export class Block {
     }
 
     // parse transactions
+    const common = Object.create(this._common)
+    const blockNumber = parseInt(this.header.number.toString('hex'), 16)
+    common.setHardforkByBlockNumber(blockNumber)
+
```
Comment (holgerd77) [inline packages/block/src/block.ts:69]: This is now [done](https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/block/src/header.ts#L146) in the header constructor (I tried to deduplicate on functionality executed in the one or the other constructor where ever possible) and therefore redundant here.

This direct setting would also undermine the new `hardforkByBlockNumber` `BlockOption` introduced in #863 . Reasoning to introduce this option is that we want to provide both possibilities for a hardfork either be kept as set (e.g. when a block is created from within the VM running on specific HF settings regardless of the block number) to have a consistent execution along with the possibility to instantiate a block which should behave regarding to the block number set (e.g. in the client context, but likely for other use cases as well).

**Comment by holgerd77** [comment]
Did a little bit of debugging here on the failing blockchain tests.

When concentrating on a running a single file with failing tests - e.g. with `ts-node ./tests/tester --blockchain --file='bcForkBlockTest'` - and injecting `console.log()` statements within `runBlock.ts` like:

```typescript
try {
    result = await applyBlock.bind(this)(block, opts)
    console.log(result)
  } catch (err) {
    console.log(err)
    await state.revert()
    throw err
  }
````
current `master` gives the following output:

```shell
$ ts-node ./tests/tester --blockchain --file='bcForkBlockTest'
+--------------------------------------------------+
| VM -> BlockchainTests                            |
|                                                  |
| TestGetterArgs                                   |
| skipTests           : 54                         |
| forkConfig          : Istanbul                   |
| file                : bcForkBlockTest            |
|                                                  |
| RunnerArgs                                       |
| forkConfigVM        : istanbul                   |
| forkConfigTestSuite : Istanbul                   |
+--------------------------------------------------+

TAP version 13
# BlockchainTests
# file: bcForkBlockTest test: BlockExtraData25_Istanbul
ok 1 correct pre stateRoot
ok 2 correct genesis RLP
Error: invalid amount of extra data
    at BlockHeader.<anonymous> (/ethereumjs-vm/packages/block/src/header.ts:320:13)
    at step (/ethereumjs-vm/packages/block/dist/header.js:33:23)
    at Object.next (/ethereumjs-vm/packages/block/dist/header.js:14:53)
    at fulfilled (/ethereumjs-vm/packages/block/dist/header.js:5:58)
ok 3 Expected exception ExtraDataTooBig
# file: bcForkBlockTest test: BlockWrongResetGas_Istanbul
ok 4 correct pre stateRoot
ok 5 correct genesis RLP
Error: Invalid Signature at tx 0
    at Block.<anonymous> (/ethereumjs-vm/packages/block/src/block.ts:175:13)
    at step (/ethereumjs-vm/packages/block/dist/block.js:33:23)
    at Object.next (/ethereumjs-vm/packages/block/dist/block.js:14:53)
    at fulfilled (/ethereumjs-vm/packages/block/dist/block.js:5:58)
    at runNextTicks (internal/process/task_queues.js:59:5)
    at processImmediate (internal/timers.js:412:9)
    at process.topLevelDomainCallback (domain.js:130:23)
ok 6 Expected exception InvalidTransactionVRS
# file: bcForkBlockTest test: BlockWrongStoreClears_Istanbul
ok 7 correct pre stateRoot
ok 8 correct genesis RLP
{
  bloom: Bloom {
    bitvector: <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 206 more bytes>
  },
  gasUsed: <BN: 59e7>,
  receiptRoot: <Buffer 6c 8a b3 6e c0 62 9c 97 73 4e 8a c8 23 cd d8 39 7d e6 7e fb 76 c7 be b9 83 be 73 dc d3 c7 81 41>,
  receipts: [
    {
      status: 1,
      gasUsed: <Buffer 59 e7>,
      bitvector: <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 206 more bytes>,
      logs: []
    }
  ],
  results: [
    {
      gasUsed: <BN: 59e7>,
      execResult: [Object],
      bloom: [Bloom],
      amountSpent: <BN: 38306>
    }
  ]
}
ok 9 Expected exception InvalidGasUsed
# file: bcForkBlockTest test: BlockWrongStoreSetGas_Istanbul
ok 10 correct pre stateRoot
ok 11 correct genesis RLP
Error: Invalid Signature at tx 0
    at Block.<anonymous> (/ethereumjs-vm/packages/block/src/block.ts:175:13)
    at step (/ethereumjs-vm/packages/block/dist/block.js:33:23)
    at Object.next (/ethereumjs-vm/packages/block/dist/block.js:14:53)
    at fulfilled (/ethereumjs-vm/packages/block/dist/block.js:5:58)
    at runNextTicks (internal/process/task_queues.js:59:5)
    at processImmediate (internal/timers.js:412:9)
    at process.topLevelDomainCallback (domain.js:130:23)
ok 12 Expected exception InvalidTransactionVRS
# file: bcForkBlockTest test: SimpleTxCosts20000_Istanbul
ok 13 correct pre stateRoot
ok 14 correct genesis RLP
Error: Invalid Signature at tx 0
    at Block.<anonymous> (/ethereumjs-vm/packages/block/src/block.ts:175:13)
    at step (/ethereumjs-vm/packages/block/dist/block.js:33:23)
    at Object.next (/ethereumjs-vm/packages/block/dist/block.js:14:53)
    at fulfilled (/ethereumjs-vm/packages/block/dist/block.js:5:58)
    at runNextTicks (internal/process/task_queues.js:59:5)
    at processImmediate (internal/timers.js:412:9)
    at process.topLevelDomainCallback (domain.js:130:23)
ok 15 Expected exception InvalidTransactionVRS

1..15
# tests 15
# pass  15
```

While running the same on the `tx-refactor` branch gives the following:

```shell
$ ts-node ./tests/tester --blockchain --file='bcForkBlockTest'
+--------------------------------------------------+
| VM -> BlockchainTests                            |
|                                                  |
| TestGetterArgs                                   |
| skipTests           : 54                         |
| forkConfig          : Istanbul                   |
| file                : bcForkBlockTest            |
|                                                  |
| RunnerArgs                                       |
| forkConfigVM        : istanbul                   |
| forkConfigTestSuite : Istanbul                   |
+--------------------------------------------------+

TAP version 13
# BlockchainTests
# file: bcForkBlockTest test: BlockExtraData25_Istanbul
ok 1 correct pre stateRoot
ok 2 correct genesis RLP
ok 3 Expected exception ExtraDataTooBig
# file: bcForkBlockTest test: BlockWrongResetGas_Istanbul
ok 4 correct pre stateRoot
ok 5 correct genesis RLP
{
  bloom: Bloom {
    bitvector: <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 206 more bytes>
  },
  gasUsed: <BN: 6596>,
  receiptRoot: <Buffer ab 6d 3e 30 83 dd ef b7 cd 46 3e 12 70 c0 7a 12 6b 51 ab 73 6a 8d 8d cc 95 a9 10 a7 3e ce 71 e9>,
  receipts: [
    {
      status: 1,
      gasUsed: <Buffer 65 96>,
      bitvector: <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 206 more bytes>,
      logs: []
    }
  ],
  results: [
    {
      gasUsed: <BN: 6596>,
      execResult: [Object],
      bloom: [Bloom],
      amountSpent: <BN: 3f7dc>
    }
  ]
}
not ok 6 last block hash
  ---
    operator: equal
    expected: |-
      '8d3d0d122068250849fce79eb37f1e6838887d298d443e5b40174333b83406ec'
    actual: |-
      '78aedd8a1a7eeba4823317a7141e37bf8c5d6369e239e82ab674c33286076f05'
    at: runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:139:11)
    stack: |-
      Error: last block hash
          at Test.assert [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:228:54)
          at Test.bound [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at Test.equal (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:389:10)
          at Test.bound (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:139:11)
          at /ethereumjs-vm-tx-refactor/packages/vm/tests/tester.js:112:13
          at fileCallback (/ethereumjs-vm-tx-refactor/packages/vm/tests/testLoader.js:56:11)
  ...
not ok 7 correct header block
  ---
    operator: equal
    expected: |-
      '8d3d0d122068250849fce79eb37f1e6838887d298d443e5b40174333b83406ec'
    actual: |-
      '78aedd8a1a7eeba4823317a7141e37bf8c5d6369e239e82ab674c33286076f05'
    at: runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:156:11)
    stack: |-
      Error: correct header block
          at Test.assert [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:228:54)
          at Test.bound [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at Test.equal (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:389:10)
          at Test.bound (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:156:11)
          at /ethereumjs-vm-tx-refactor/packages/vm/tests/tester.js:112:13
          at fileCallback (/ethereumjs-vm-tx-refactor/packages/vm/tests/testLoader.js:56:11)
  ...
# file: bcForkBlockTest test: BlockWrongStoreClears_Istanbul
ok 8 correct pre stateRoot
ok 9 correct genesis RLP
{
  bloom: Bloom {
    bitvector: <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 206 more bytes>
  },
  gasUsed: <BN: 59e7>,
  receiptRoot: <Buffer 6c 8a b3 6e c0 62 9c 97 73 4e 8a c8 23 cd d8 39 7d e6 7e fb 76 c7 be b9 83 be 73 dc d3 c7 81 41>,
  receipts: [
    {
      status: 1,
      gasUsed: <Buffer 59 e7>,
      bitvector: <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 206 more bytes>,
      logs: []
    }
  ],
  results: [
    {
      gasUsed: <BN: 59e7>,
      execResult: [Object],
      bloom: [Bloom],
      amountSpent: <BN: 38306>
    }
  ]
}
not ok 10 last block hash
  ---
    operator: equal
    expected: |-
      'fe68e7d76854990d722d204f2eccc1a8c10a730deb7816cf28bbb7f3dbbcaacf'
    actual: |-
      '22f5ba1ed9ba7a09b9bc8d6f8c4a182c30e4219c775d47e2bd3f40ba5bcc260d'
    at: runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:139:11)
    stack: |-
      Error: last block hash
          at Test.assert [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:228:54)
          at Test.bound [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at Test.equal (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:389:10)
          at Test.bound (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:139:11)
          at /ethereumjs-vm-tx-refactor/packages/vm/tests/tester.js:112:13
          at fileCallback (/ethereumjs-vm-tx-refactor/packages/vm/tests/testLoader.js:56:11)
  ...
not ok 11 correct header block
  ---
    operator: equal
    expected: |-
      'fe68e7d76854990d722d204f2eccc1a8c10a730deb7816cf28bbb7f3dbbcaacf'
    actual: |-
      '22f5ba1ed9ba7a09b9bc8d6f8c4a182c30e4219c775d47e2bd3f40ba5bcc260d'
    at: runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:156:11)
    stack: |-
      Error: correct header block
          at Test.assert [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:228:54)
          at Test.bound [as _assert] (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at Test.equal (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:389:10)
          at Test.bound (/ethereumjs-vm-tx-refactor/node_modules/tape/lib/test.js:80:32)
          at runBlockchainTest (/ethereumjs-vm-tx-refactor/packages/vm/tests/BlockchainTestsRunner.js:156:11)
          at /ethereumjs-vm-tx-refactor/packages/vm/tests/tester.js:112:13
          at fileCallback (/ethereumjs-vm-tx-refactor/packages/vm/tests/testLoader.js:56:11)
  ...

// I'll cut here a bit due to getting too bulky

1..19
# tests 19
# pass  11
# fail  8
```

So when looking at the expected exceptions `InvalidTransactionVRS` along with the errors thrown on `master` (`Error: Invalid Signature at tx 0`) it seems that transactions in the new tx class are actually "fixed" in a way they shouldn't be and they should actually throw/fail?

Just an on-first-sight interpretation, it might also be the case that the fixes here are correct and our test setup is wrong, this is less likely though I would say. The work from @jochem-brouwer on #849 should help to make such kind of analysis less disambiguous by running the tests to the end (these early stops on test runs are also the reason for the differing in tests executed on the comparison from above should - I would assume - be unrelated to the root of the problem though.

File: packages/block/src/block.ts (hunk 10 above)
Comment (ryanio) [reply packages/block/src/block.ts]: resolved in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/bc0e0a8be2c8adfb78e926cbc140dc8985491611

**Comment by jochem-brouwer** [comment]
@holgerd77 FYI that injection of code is semi-equivalent to hooking up to the `afterBlock` event, so you don't have to edit VMs code: `vm.on('afterBlock', console.log)`

Just glanced over these blockchain tests, very confusing, these have to do with block tests 🤔

File: packages/tx/src/transaction.ts [hunk 11]
```diff
@@ -21,290 +21,255 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
...
+  public readonly gasLimit: BN
+  public readonly gasPrice: BN
+  public readonly to?: Address
+  public readonly value: BN
+  public readonly data: Buffer
+  public readonly v?: BN
+  public readonly r?: BN
+  public readonly s?: BN
+
+  public static fromTxData(txData: TxData, common?: Common) {
```
Comment (holgerd77) [inline packages/tx/src/transaction.ts:35]: Just realizing while working on a similar transition on the block library: do we want to put `common` here in an options dict? Think this would be more future-proof, will also introduce the like on block since there are options which needed to be passed around.

Comment (ryanio) [reply packages/tx/src/transaction.ts:35]: hmm I liked the simplicity of removing the need for TxOptions, but i see your point for future-proofing.

i assume we would do it for the constructor and all the factory methods?

File: packages/vm/lib/runBlockchain.ts [hunk 12]
```diff
@@ -29,6 +29,7 @@ export default async function runBlockchain(this: VM, blockchain?: Blockchain) {
     // run block, update head if valid
     try {
       await this.runBlock({ block, root: parentState })
+      console.log()
```
Comment (holgerd77) [inline packages/vm/lib/runBlockchain.ts:32]: Ug. 😛

File: packages/tx/src/transaction.ts (hunk 11 above)
Comment (holgerd77) [reply packages/tx/src/transaction.ts:35]: Yes, would say so, please also put TxOptions being the last parameter for the constructor for consistency (unless you have got an explicit reason not to).

Yeah, sorry for un-simplify here a bit, but I think this is quite a realistic scenario that we want to add here at some point and it would be a bit a pity if this would be blocked by not have this prepared.

File: packages/tx/src/transaction.ts [hunk 13]
```diff
@@ -21,340 +21,373 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
...
-
-    return !!this._senderPubKey
   }
 
   /**
-   * sign a transaction with a given private key
+   * Sign a transaction with a given private key
    * @param privateKey - Must be 32 bytes in length
    */
   sign(privateKey: Buffer) {
```
Comment (holgerd77) [inline packages/tx/src/transaction.ts:220]: I think we should point this out more clearly in the code docs that this only gives a signed transaction if assigned to a new variable, otherwise this is too dangerous that people assume their transaction to be signed when just using `tx.sign()` (and not further thinking about immutability, which is in fact rather likely than unlikely I would assume).

**Review by holgerd77 — APPROVED** [review]
LGTM

Some last comments for the next cosmetic finalization round. 😄 

Will merge.

**Comment by holgerd77** [comment]
🎉

File: packages/vm/lib/runBlockchain.ts (hunk 12 above)
Comment (ryanio) [reply packages/vm/lib/runBlockchain.ts:32]: boo, i'll clean this up :)
